connector/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetRangeCalculator.scala

```diff
@@ -47,13 +48,15 @@ private[kafka010] class KafkaOffsetRangeCalculator(val minPartitions: Option[Int
     val offsetRanges = ranges.filter(_.size > 0)
 
     // If minPartitions not set or there are enough partitions to satisfy minPartitions
-    if (minPartitions.isEmpty || offsetRanges.size >= minPartitions.get) {
+    // and maxBytesPerPartition is empty
+    if ((minPartitions.isEmpty || offsetRanges.size >= minPartitions.get)
+        && maxRecordsPerPartition.isEmpty) {
       // Assign preferred executor locations to each range such that the same topic-partition is
       // preferentially read from the same executor and the KafkaConsumer can be reused.
       offsetRanges.map { range =>
         range.copy(preferredLoc = getLocation(range.topicPartition, executorLocations))
       }
-    } else {
+    } else if (minPartitions.isDefined && minPartitions.get > offsetRanges.size) {
```

Comment:
This means if `minPartitions` and `maxRecordsPerPartition` are both set by a user, the `minPartitions` supersedes it. Hence we might still end up with partitions with records greater than `maxRecordsPerPartition`. 

This behavior won't be obvious to users, who may end up setting both and may think they work together to make sure we have at least `minPartitions` partitions and no partition will have more than `maxRecordsPerPartition` records.

One solution is, after doing this minPartitions splitting, we can also call the code that ensures that no partition has more than `maxRecordsPerPartition` if specified, hence further splitting the partitions.

Reply:
Thanks for highlighting this issue. I am dividing partitions based on **maxRecordsPerPartition** first so that partitions have uniform size data, then I am passing these partitions for further split based on **minPartition** config.